File: leetcode/solution_973.py
'''
Summary - Attempt #1

Your own answer?: Yes
Time spent: 13m

Time complexity: O(NlogN)
Runtime: 948 ms, faster than 27.26% of Python3 online submissions for K Closest Points to Origin.
Space complexity: O(N)
Memory Usage: 19.8 MB, less than 61.82% of Python3 online submissions for K Closest Points to Origin.
'''

import sys
from typing import List
input = sys.stdin.readline

class Solution:
    def __init__(self) -> None:
        points = [[1,3],[-2,2]]
        k = 1
        print(self.kClosest(points, k))

    def kClosest(self, points: List[List[int]], k: int) -> List[List[int]]:
        store = []
        for point in points:
            store.append([point[0]*point[0] + point[1]*point[1], point])
        store.sort()
        
        return [item[1] for item in store[:k]]

Solution()

File: leetcode/solution_973.cs
// Time Complexity: O(nlogn) where n is the number of elements in the given array
// Space Complexity: O(n)

/*
brute-force approach would take TC O(nlogn)
*/

public class Solution {
    public int[][] KClosest(int[][] points, int k) {
        var storage = new List<(int, int)>();
        
        for (int i = 0; i < points.Length; i++)
        {
            var x = points[i][0];
            var y = points[i][1];
            var distance = x * x + y * y;
            storage.Add((distance, i));    
        }
        
        storage.Sort();
        
        var res = new int[k][];
        
        for (var i = 0; i < k; i++)
        {
            var (distance, index) = storage[i];
            res[i] = points[index];
        }
        
        return res;
    }
}